core/common/src/main/java/alluxio/network/TieredIdentityFactory.java

```diff
@@ -0,0 +1,123 @@
+import alluxio.util.ShellUtils;
+import alluxio.wire.TieredIdentity;
+import alluxio.wire.TieredIdentity.LocalityTier;
+
+import com.google.common.base.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nullable;
+
+/**
+ * Class for getting tiered identity.
+ */
+public final class TieredIdentityFactory {
+  private static TieredIdentity sInstance = null;
+
+  /**
+   * @return the singleton tiered identity instance for this JVM
+   */
+  public static TieredIdentity getInstance() {
+    synchronized (sInstance) {
+      if (sInstance == null) {
+        sInstance = create();
+      }
+      return sInstance;
+    }
+  }
+
+  /**
+   * Creates a tiered identity based on configuration.
+   *
+   * @return the created tiered identity
+   */
+  public static TieredIdentity create() {
+    TieredIdentity scriptIdentity = null;
+    // If a script is configured, run the script to get all locality tiers.
+    if (Configuration.containsKey(PropertyKey.LOCALITY_SCRIPT)) {
+      scriptIdentity = fromScript();
+    }
+
+    List<LocalityTier> tiers = new ArrayList<>();
+    List<String> orderedTierNames = Configuration.getList(PropertyKey.LOCALITY_ORDER, ",");
+    for (int i = 0; i < orderedTierNames.size(); i++) {
+      String tierName = orderedTierNames.get(i);
+      String value = null;
+      if (scriptIdentity != null) {
+        LocalityTier scriptTier = scriptIdentity.getTiers().get(i);
+        Preconditions.checkState(scriptTier.getTierName().equals(tierName));
+        value = scriptTier.getValue();
+      }
+      // Explicit configuration overrides script output.
+      if (Configuration.containsKey(Template.LOCALITY_TIER.format(tierName))) {
+        value = Configuration.get(Template.LOCALITY_TIER.format(tierName));
+      }
+      tiers.add(new LocalityTier(tierName, value));
+    }
+    return new TieredIdentity(tiers);
+  }
+
+  /**
+   * @return a tiered identity created from running the user-provided script
+   */
+  public static TieredIdentity fromScript() {
+    String script = Configuration.get(PropertyKey.LOCALITY_SCRIPT);
+    String identityString;
+    try {
+      identityString = ShellUtils.execCommand(script);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          String.format("Failed to run script %s: %s", script, e.toString(), e));
+    }
+    TieredIdentity identity = parseIdentityString(identityString);
+    if (identity == null) {
+      throw new RuntimeException(String.format(
+          "Failed to parse the output of running %s. "
+              + "The value should be a comma-separated list of key=value pairs, but was %s",
+          script, identityString));
+    }
+    return identity;
+  }
+
+  /**
+   * @param identityString tiered identity string to parse
+   * @return the parsed tiered identity, or null if the string could not be parsed
```

Comment:
Should we just throw runtime exception here?

Reply:
done